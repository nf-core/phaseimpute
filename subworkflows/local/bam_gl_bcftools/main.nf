include { GAWK                      } from '../../../modules/nf-core/gawk'
include { TABIX_BGZIP               } from '../../../modules/nf-core/tabix/bgzip'
include { BCFTOOLS_MPILEUP          } from '../../../modules/nf-core/bcftools/mpileup'
include { BCFTOOLS_ANNOTATE         } from '../../../modules/nf-core/bcftools/annotate'

workflow BAM_GL_BCFTOOLS {

    take:
    ch_input   // channel: [ [id], bam, bai ]
    ch_posfile // channel: [ [panel, chr], legend]
    ch_fasta   // channel: [ [genome], fasta, fai]

    main:

    ch_versions      = Channel.empty()
    ch_multiqc_files = Channel.empty()

    // Convert legend to TSV with ','
    GAWK(ch_posfile, [])
    ch_versions = ch_versions.mix(GAWK.out.versions)

    // Compress TSV
    TABIX_BGZIP(GAWK.out.output)
    ch_versions = ch_versions.mix(TABIX_BGZIP.out.versions.first())

    ch_mpileup       = ch_input
        .combine(TABIX_BGZIP.out.output)
        .map{metaI, bam, bai, metaPC, tsv ->
                [metaI + ["panel": metaPC.id, "chr": metaPC.chr], bam, tsv]
        }

    BCFTOOLS_MPILEUP(
        ch_mpileup,
        ch_fasta,
        false
    )
    ch_versions = ch_versions.mix(BCFTOOLS_MPILEUP.out.versions.first())

    // Annotate the variants
    BCFTOOLS_ANNOTATE(BCFTOOLS_MPILEUP.out.vcf
        .join(BCFTOOLS_MPILEUP.out.tbi)
        .combine(Channel.of([[], [], [], []]))
    )
    ch_versions = ch_versions.mix(BCFTOOLS_ANNOTATE.out.versions.first())

    // Output
    ch_output = BCFTOOLS_ANNOTATE.out.vcf
        .join(BCFTOOLS_ANNOTATE.out.tbi)

    ch_multiqc_files = ch_multiqc_files.mix(BCFTOOLS_MPILEUP.out.stats.map{ it[1] })

    emit:
    vcf           = ch_output        // channel: [ [id, panel, chr], vcf, tbi ]
    versions      = ch_versions      // channel: [ versions.yml ]
    multiqc_files = ch_multiqc_files
}
