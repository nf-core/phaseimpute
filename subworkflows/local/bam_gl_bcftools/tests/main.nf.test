nextflow_workflow {

    name "Test Subworkflow BAM_GL_BCFTOOLS"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_GL_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_gl_bcftools"
    tag "bam_gl_bcftools"

    tag "gawk"
    tag "tabix"
    tag "tabix/bgzip"
    tag "bcftools"
    tag "bcftools/mpileup"
    tag "bcftools/annotate"

    test("Compute genotype likelihood") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf.collect{ path(it[1]).vcf.summary }).match("Summary") }
            )
        }
    }
}
