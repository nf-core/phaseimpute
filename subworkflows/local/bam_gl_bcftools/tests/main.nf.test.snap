{
    "Summary": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:33:11.627895132"
    },
    "Compute genotype likelihood": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22"
                        },
                        "NA12878.annotate.vcf.gz:md5,d3b585236be8904742a8635d5623daec",
                        "NA12878.annotate.vcf.gz.tbi:md5,dbc7f7bd9a5e9b299cd2577b05061897"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22"
                        },
                        "NA19401.annotate.vcf.gz:md5,fbae0f875f9351f7a1df28953b99f17b",
                        "NA19401.annotate.vcf.gz.tbi:md5,7ca03da29e0bcc4f34b9842069a5ec2c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3779c6b3c000bc757d3d6109010518b0",
                    "versions.yml:md5,4739962c934731254db8b31083aa0df5",
                    "versions.yml:md5,544ff5e39aa23f18c41587a739beaf34",
                    "versions.yml:md5,c9d32dd20061f19422ff4d4b6f3fb24d"
                ],
                "2": [
                    "NA12878.bcftools_stats.txt:md5,4cacbe4d6dfba4a8035e37611c91923c",
                    "NA19401.bcftools_stats.txt:md5,f91fd586acfdf3845d111a062fbb263f"
                ],
                "multiqc_files": [
                    "NA12878.bcftools_stats.txt:md5,4cacbe4d6dfba4a8035e37611c91923c",
                    "NA19401.bcftools_stats.txt:md5,f91fd586acfdf3845d111a062fbb263f"
                ],
                "vcf": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22"
                        },
                        "NA12878.annotate.vcf.gz:md5,d3b585236be8904742a8635d5623daec",
                        "NA12878.annotate.vcf.gz.tbi:md5,dbc7f7bd9a5e9b299cd2577b05061897"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22"
                        },
                        "NA19401.annotate.vcf.gz:md5,fbae0f875f9351f7a1df28953b99f17b",
                        "NA19401.annotate.vcf.gz.tbi:md5,7ca03da29e0bcc4f34b9842069a5ec2c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3779c6b3c000bc757d3d6109010518b0",
                    "versions.yml:md5,4739962c934731254db8b31083aa0df5",
                    "versions.yml:md5,544ff5e39aa23f18c41587a739beaf34",
                    "versions.yml:md5,c9d32dd20061f19422ff4d4b6f3fb24d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:33:11.067033683"
    }
}